### PR TITLE
design pattern assigment added factory

### PR DESCRIPTION
**Pull request by mayuriv-webonise**

### CODE REVIEW

File: assignment3.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+# Subclasses for different divisions
+class DevelopmentEmployee < Employee; end
+class QAEmployee < Employee; end
+class DevopsEmployee < Employee; end
+
+class EmployeeFactory
+  def create_employee(name, age, department, division, designation)
+    case division
+    when 'Development'
+      DevelopmentEmployee.new(name, age, department, division, designation)
```
Comment (venkatesh-webonise) [inline assignment3.rb]: if you are already checking with division, no need to pass division to the constructor while initializing.

File: assignment3.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    @designation = designation
+  end
+
+  def to_csv_row
+    [name, age, department, division, designation]
+  end
+end
+
+# Subclasses for different divisions
+class DevelopmentEmployee < Employee; end
```
Comment (venkatesh-webonise) [inline assignment3.rb]: please remove Employee in child classes.

File: assignment3.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    CSV.open('employees.csv', 'w') do |csv|
+      csv << ['Name', 'Age', 'Department', 'Division', 'Designation']
+      @employees.each do |employee|
+        csv << employee.to_csv_row
+      end
+    end
+  end
+end
+
+class Department
```
Comment (venkatesh-webonise) [inline assignment3.rb]: Each class should be in a separate file.

File: assignment3.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
+  require 'csv'
+  require("./employee.rb")
+  require("./factory_method.rb")
+  require("./hr_class.rb")
```
Comment (venkatesh-webonise) [inline assignment3.rb]: please change file name hr_class to hr and also for other files.

File: department_class.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+  require("./assignment3.rb")
```
Comment (venkatesh-webonise) [inline department_class.rb]: I don't think you need to require assignment3.rb file here.

File: employee.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+  require("./assignment3.rb")
```
Comment (venkatesh-webonise) [inline employee.rb]: I don't think you need to require assignment3.rb file here.

File: factory_method.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+  require("./assignment3.rb")
```
Comment (venkatesh-webonise) [inline factory_method.rb]: I don't think you need to require assignment3.rb file here.

File: employee.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
+require("./assignment3.rb")
+  class Employee
```
Comment (venkatesh-webonise) [inline employee.rb]: please check alignment

File: employee.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
+require("./assignment3.rb")
```
Comment (venkatesh-webonise) [inline employee.rb]: this line is not needed I guess.

Comment (mayuriv-webonise) [reply employee.rb]: We require this as it will not show any output

Comment (mayuriv-webonise) [reply employee.rb]: removed

File: employee.rb (hunk 8 above)
Comment (mayuriv-webonise) [reply employee.rb]: added alignments

File: factory_method.rb (hunk 7 above)
Comment (mayuriv-webonise) [reply factory_method.rb]: Yes, removed require

File: employee.rb (hunk 6 above)
Comment (mayuriv-webonise) [reply employee.rb]: Yes, removed require

File: department_class.rb (hunk 5 above)
Comment (mayuriv-webonise) [reply department_class.rb]: Yes, removed require

File: assignment3.rb (hunk 4 above)
Comment (mayuriv-webonise) [reply assignment3.rb]: changes done

File: assignment3.rb (hunk 3 above)
Comment (mayuriv-webonise) [reply assignment3.rb]: done chnges

File: assignment3.rb (hunk 2 above)
Comment (mayuriv-webonise) [reply assignment3.rb]: removed employee

File: assignment3.rb (hunk 1 above)
Comment (mayuriv-webonise) [reply assignment3.rb]: removed devision

**Review by venkatesh-webonise — APPROVED** [review]

**Pull request merged**